src/singlesqlitebackend.cpp

```diff
@@ -0,0 +1,897 @@
+    }
+
+    d->mAdded.clear();
+    d->mModified.clear();
+    d->mDeleted.clear();
+
+    return false;
+}
+
+bool SingleSqliteBackend::purgeDeletedIncidences(const QString &notebookUid,
+                                                 const Incidence::List &list)
+{
+    if (!d->mFormat) {
+        return false;
+    }
+
+    if (!d->mSem.acquire()) {
+        qCWarning(lcMkcal) << "cannot lock" << d->mDatabaseName << "error" << d->mSem.errorString();
+        return false;
+    }
+
+    int rv = 0;
+    unsigned int error = 1;
+
+    char *errmsg = NULL;
+    const char *query = NULL;
+
+    query = BEGIN_TRANSACTION;
+    SL3_exec(d->mFormat->database());
+
+    error = 0;
+    for (const Incidence::Ptr &incidence: list) {
+        if (!d->mFormat->purgeDeletedComponents(*incidence, notebookUid)) {
+            error += 1;
+        }
+    }
+
+    query = COMMIT_TRANSACTION;
+    SL3_exec(d->mFormat->database());
+
+ error:
+    if (!d->mSem.release()) {
+        qCWarning(lcMkcal) << "cannot release lock" << d->mDatabaseName << "error" << d->mSem.errorString();
+    }
+    return error == 0;
+}
+
+bool SingleSqliteBackend::Private::loadNotebooks(Notebook::List *list, Notebook::Ptr *defaultNb, sqlite3_stmt *stmt)
+{
+    if (!mSem.acquire()) {
+        qCWarning(lcMkcal) << "cannot lock" << mDatabaseName << "error" << mSem.errorString();
+        return false;
+    }
+
+    bool isDefault;
+    Notebook::Ptr nb;
+    while ((nb = mFormat->selectCalendars(stmt, &isDefault))) {
+        qCDebug(lcMkcal) << "loaded notebook" << nb->uid() << nb->name() << "from database";
+        list->append(nb);
+        if (isDefault && defaultNb) {
+            *defaultNb = nb;
+        }
+    }
+
+    if (!mSem.release()) {
+        qCWarning(lcMkcal) << "cannot release lock" << mDatabaseName << "error" << mSem.errorString();
+    }
+
+    return true;
+}
+
+bool SingleSqliteBackend::notebooks(Notebook::List *list, Notebook::Ptr *defaultNb)
+{
+    int rv = 0;
+    sqlite3_stmt *stmt = NULL;
+    bool success = false;
+
+    if (!d->mFormat || !list) {
+        return false;
+    }
+
+    const char *query = SELECT_CALENDARS_ALL;
+    int qsize = sizeof(SELECT_CALENDARS_ALL);
+    SL3_prepare_v2(d->mFormat->database(), query, qsize, &stmt, nullptr);
+
+    success = d->loadNotebooks(list, defaultNb, stmt);
+
+error:
+    sqlite3_finalize(stmt);
+    return success;
+}
+
+bool SingleSqliteBackend::Private::saveNotebook(const Notebook &nb, DBOperation dbop,
+                                                sqlite3_stmt *stmt, bool isDefault)
+{
+    bool success;
+    const char *operation = (dbop == DBInsert) ? "inserting" :
+                            (dbop == DBUpdate) ? "updating" : "deleting";
+
+    if (!mSem.acquire()) {
+        qCWarning(lcMkcal) << "cannot lock" << mDatabaseName << "error" << mSem.errorString();
+        return false;
+    }
+
+    if ((success = mFormat->modifyCalendars(nb, dbop, stmt, isDefault))) {
```

Comment:
Side-note: strange existing naming on SqliteFormat on modifyCalendars() when it's a single calendar and other api talks about notebooks. In similar way selection and modifyComponents() etc methods.